src/out_ref.rs

```diff
@@ -0,0 +1,573 @@
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut T> for Out<'out, T>
+where
+    T : Copy, // prevent accidentally leaking memory
+{
+    #[inline]
+    fn from (p: &'out mut T)
+      -> Out<'out, T>
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - `T` and `MaybeUninit<T>` have the same layout
+            //     (`#[repr(transparent)]`)
+            //
+            //   - The API does not allow to write `MaybeUninit::uninit()` into
+            //     the pointee.
+            mem::transmute::<_, &'out mut MaybeUninit<T>>(p).into()
+        }
+    }
+}
+
+/// For non-`Copy` types, explicitely transmuting the `mut` reference into one
+/// that points to a `ManuallyDrop` is required, so as to express how likely it
+/// is that memory be leaked.
+impl<'out, T : 'out> From<&'out mut ManuallyDrop<T>> for Out<'out, T> {
+    #[inline]
+    fn from (p: &'out mut ManuallyDrop<T>)
+      -> Out<'out, T>
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - `ManuallyDrop<T>` and `MaybeUninit<T>` have the same layout
+            //     (`#[repr(transparent)]`)
+            //
+            //   - The API does not allow to write `MaybeUninit::uninit()` into
+            //     the pointee.
+            mem::transmute::<_, &'out mut MaybeUninit<T>>(p).into()
+        }
+    }
+}
+
+impl<'out, T : 'out + ?Sized> Out<'out, T> {
+    #[inline]
+    pub
+    fn new (p: impl Into<Out<'out, T>>)
+      -> Self
+    {
+        p.into()
+    }
+
+    #[inline]
+    pub
+    fn reborrow<'reborrow> (self: &'reborrow mut Out<'out, T>)
+      -> Out<'reborrow, T>
+    where
+        'out : 'reborrow,
+    {
+        Out(self.0, Default::default())
+    }
+
+    /// Shorthand for [`.reborrow()`][`Out::reborrow`]
+    #[inline]
+    pub
+    fn r<'reborrow> (self: &'reborrow mut Out<'out, T>)
+      -> Out<'reborrow, T>
+    where
+        'out : 'reborrow,
+    {
+        self.reborrow()
+    }
+}
+
+impl<'out, T : 'out> Out<'out, T> {
+    #[inline]
+    pub
+    fn write (self: Out<'out, T>, value: T)
+      -> &'out mut T
+    {
+        unsafe {
+            // Safety: this writes a valid (non garbage) value to the pointee
+            self.0.as_ptr().write(value);
+            self.assume_init()
+        }
+    }
+
+    #[inline]
+    pub
+    fn replace (self: Out<'out, T>, value: T)
+      -> (MaybeUninit<T>, &'out mut T)
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - This:
+            //
+            //      1. writes a valid value to the pointee,
+            //
+            //      2. extracts the previous value as now owned `MaybeUninit<T>`
+            //
+            //     Thus it does not not `assume_init()` the pointee nor write
+            //     garbage to it.
+            //
+            //   - Thanks to write the pointee is known to be initialized.
+            (
+                ptr::replace(self.0.as_ptr().cast(), MaybeUninit::new(value)),
```

Comment:
`mem::replace(self.as_uninit_mut(), MaybeUninit::new(value))`

I think that makes the proof obligations clearer.

Reply:
True, I wrote that function before the `as_uninit_mut` helper existed 👍